libraries/AP_BattMonitor/AP_BattMonitor_Scripting.cpp

```diff
@@ -0,0 +1,87 @@
+#include "AP_BattMonitor_Scripting.h"
+
+#if AP_BATTERY_SCRIPTING_ENABLED
+
+#define AP_BATT_MONITOR_SCRIPTING_TIMEOUT_US 5000000
+
+bool AP_BattMonitor_Scripting::capacity_remaining_pct(uint8_t &percentage) const
+{
+    if (internal_state.capacity_remaining_pct != UINT8_MAX) {
+        percentage = internal_state.capacity_remaining_pct;
+        return true;
+    }
+    // Fall back to default implementation
+    return AP_BattMonitor_Backend::capacity_remaining_pct(percentage);
+}
+
+bool AP_BattMonitor_Scripting::get_cycle_count(uint16_t &cycles) const
+{
+    if (internal_state.cycle_count == UINT8_MAX) {
+        return false;
+    }
+    cycles = internal_state.cycle_count;
+    return true;
+}
+
+// Called by frontend to update the state. Called at 10Hz
+void AP_BattMonitor_Scripting::read()
+{
+    WITH_SEMAPHORE(sem);
+
+    // Check for timeout, to prevent a faulty script from appearing healthy
+    if (last_update_us == 0 || AP_HAL::micros() - last_update_us > AP_BATT_MONITOR_SCRIPTING_TIMEOUT_US) {
+        _state.healthy = false;
+        return;
+    }
+
+    if (_state.last_time_micros == last_update_us) {
+        // No new data
+        return;
+    }
+
+    /*
+      the script can fill in voltages up to 32 cells, for mavlink reporting
+      the extra cell voltages get distributed over the max of 14 for mavlink
+     */
+    for (uint8_t i = 0; i < MIN(AP_BATT_MONITOR_CELLS_MAX,internal_state.cell_count); i++) {
+        _state.cell_voltages.cells[i] = internal_state.cell_voltages[i];
+    }
+    _state.voltage = internal_state.voltage;
+    const bool has_current = !isnan(internal_state.current_amps);
+    if (has_current) {
```

Comment:
```suggestion
    if (!isnan(internal_state.current_amps)) {
```
This matches with the other checks in this function.

Reply:
changed